update-engine/src/events.rs

```diff
@@ -0,0 +1,688 @@
+
+        /// The amount of time this attempt has taken so far.
+        attempt_elapsed: Duration,
+
+        /// A message assocaited with the reset.
+        message: Cow<'static, str>,
+    },
+
+    /// An attempt failed and this step is being retried.
+    AttemptRetry {
+        /// Information about the step.
+        step: StepInfoWithMetadata<S>,
+
+        /// The attempt number for the next attempt.
+        next_attempt: usize,
+
+        /// Total time elapsed since the start of the step. Includes prior
+        /// attempts.
+        step_elapsed: Duration,
+
+        /// The amount of time the previous attempt took.
+        attempt_elapsed: Duration,
+
+        /// A message associated with the retry.
+        message: Cow<'static, str>,
+    },
+
+    /// A step is complete and the next step has been started.
+    StepCompleted {
+        /// Information about the step that just completed.
+        step: StepInfoWithMetadata<S>,
+
+        /// The attempt number that completed.
+        attempt: usize,
+
+        /// The outcome of the step.
+        outcome: StepOutcome<S>,
+
+        /// The next step that is being started.
+        next_step: StepInfoWithMetadata<S>,
+
+        /// Total time elapsed since the start of the step. Includes prior
+        /// attempts.
+        step_elapsed: Duration,
+
+        /// The time it took for this attempt to complete.
+        attempt_elapsed: Duration,
+    },
+
+    /// Execution is complete.
+    ///
+    /// This is a terminal event: it is guaranteed that no more events will be
+    /// seen after this one.
+    ExecutionCompleted {
+        /// Information about the last step that completed.
+        last_step: StepInfoWithMetadata<S>,
+
+        /// The attempt number that completed.
+        last_attempt: usize,
+
+        /// The outcome of the last step.
+        last_outcome: StepOutcome<S>,
+
+        /// Total time elapsed since the start of the step. Includes prior
+        /// attempts.
+        step_elapsed: Duration,
+
+        /// The time it took for this attempt to complete.
+        attempt_elapsed: Duration,
+    },
+
+    /// Execution failed.
+    ///
+    /// This is a terminal event: it is guaranteed that no more events will be
+    /// seen after this one.
+    ExecutionFailed {
+        /// Information about the step that failed.
+        failed_step: StepInfoWithMetadata<S>,
+
+        /// The total number of attempts that were performed before the step failed.
+        total_attempts: usize,
+
+        /// Total time elapsed since the start of the step. Includes prior
+        /// attempts.
+        step_elapsed: Duration,
+
+        /// The time it took for this attempt to complete.
+        attempt_elapsed: Duration,
+
+        /// A message associated with the failure.
+        message: String,
+
+        /// A chain of causes associated with the failure.
+        causes: Vec<String>,
+    },
+
+    /// Future variants that might be unknown.
```

Comment:
This feels almost overly defensive to me, but I'm not sure why, exactly. Most internal API enums do not have this kind of catch-all case, I guess?

Reply:
My main concern was cross-version compatibility with wicketd and installinator -- if a new event gets added to the update engine and to the installinator, we want to ensure that wicketd on the old version doesn't fail while deserializing events from the installinator.

Reply:
Right, that makes sense. I think my (admittedly vague) feeling is that _lots_ of internal API enums will currently fail if one side adds a variant that it uses before the other side has been updated. Do we need to add this kind of future-variant-unknown handling to everything? Or can we rely on the update system to keep things in sync when we have API changes?

Reply:
I think we need to add unknown variant handling to anything that isn't monolithically versioned. (For example, that's why the separation between `ArtifactKind` and `KnownArtifactKind` exists.)